contrib/gorm.io/gorm.v1/gorm.go

```diff
@@ -0,0 +1,123 @@
+)
+
+type key string
+
+const (
+	gormSpanStartTimeKey = key("dd-trace-go:span")
+)
+
+// Open opens a new (traced) database connection. The used driver must be formerly registered
+// using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
+func Open(dialector gorm.Dialector, cfg *gorm.Config, opts ...Option) (*gorm.DB, error) {
+	db, err := gorm.Open(dialector, cfg)
+	if err != nil {
+		return db, err
+	}
+	return withCallbacks(db, opts...)
+}
+
+func withCallbacks(db *gorm.DB, opts ...Option) (*gorm.DB, error) {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+
+	afterFunc := func(operationName string) func(*gorm.DB) {
+		return func(db *gorm.DB) {
+			after(db, operationName, cfg)
+		}
+	}
+
+	cb := db.Callback()
+	err := cb.Create().Before("gorm:create").Register("dd-trace-go:before_create", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Create().After("gorm:create").Register("dd-trace-go:after_create", afterFunc("gorm.create"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Update().Before("gorm:update").Register("dd-trace-go:before_update", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Update().After("gorm:update").Register("dd-trace-go:after_update", afterFunc("gorm.update"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Delete().Before("gorm:delete").Register("dd-trace-go:before_delete", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Delete().After("gorm:delete").Register("dd-trace-go:after_delete", afterFunc("gorm.delete"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Query().Before("gorm:query").Register("dd-trace-go:before_query", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Query().After("gorm:query").Register("dd-trace-go:after_query", afterFunc("gorm.query"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Row().Before("gorm:query").Register("dd-trace-go:before_row_query", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Row().After("gorm:query").Register("dd-trace-go:after_row_query", afterFunc("gorm.row_query"))
+	if err != nil {
+		return db, err
+	}
+	return db, nil
+}
+
+func before(scope *gorm.DB) {
+	if scope.Statement != nil && scope.Statement.Context != nil {
+		scope.Statement.Context = context.WithValue(scope.Statement.Context, gormSpanStartTimeKey, time.Now())
+	}
+}
+
+func after(db *gorm.DB, operationName string, cfg *config) {
+	if db.Statement == nil || db.Statement.Context == nil {
+		return
+	}
+
+	ctx := db.Statement.Context
+	t, ok := ctx.Value(gormSpanStartTimeKey).(time.Time)
+	if !ok {
+		return
+	}
+
+	opts := []ddtrace.StartSpanOption{
+		tracer.StartTime(t),
+		tracer.ServiceName(cfg.serviceName),
+		tracer.SpanType(ext.SpanTypeSQL),
+		tracer.ResourceName(db.Statement.SQL.String()),
+	}
+	if !math.IsNaN(cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+	}
+
+	span, _ := tracer.StartSpanFromContext(ctx, operationName, opts...)
```

Comment:
If I understand correctly, this creates a span after an operation has finished. Doesn't this mean that any child spans created during the operation (e.g., for the raw database queries via `database/sql`) will have the wrong parent span?

Reply:
It will be fine, the idea here is that the span has the `tracer.StartTime` option which places the span in the past, to when the operation started.

Reply:
The span would have the correct time yes, but the context isn't updated in `before` so a child span would still have the wrong parent, no?

Reply:
Ah, right. I see what you mean. If someone were to use `db.Statement.Context`. In that case you are right, it wouldn't have _this_ parent. But the question is - would you want it to? Any span started as a child of this span would start after this one finishes, which seems unusual but if you have a valid use case for it we can consider it.

Reply:
Say you've enabled tracing for the `database/sql` connection underlying this `gorm.DB`. Then a child span here (e.g., `postgres.db SELECT * from foo`) would be created that would start and end within the gorm span but not have the gorm span as parent... I think.

Reply:
I'll let the author jump in and hopefully explain the reasoning behind this approach @AdallomRoy

Reply:
Hi @gbbr and @hatstand - Thanks for your comments.
I was trying to create this package as identical as possible to the gorm v1 package here: https://github.com/DataDog/dd-trace-go/blob/v1/contrib/gopkg.in/jinzhu/gorm.v1/gorm.go
so that if you upgrade you will have the same experience.
As far as I understand, this happens in the "after" func in that package as well. Do you think this would behave differently?

Reply:
We could definitely start the span in the before func and save it as part of the context and grab it here for finish if we think that behavior is better

Reply:
My view is that we shouldn't make any premature improvements if this code has had no problems reported so far. As long as there is no issue, let's not fix it 🙂

@hatstand please feel free to open up an issue and describe how to reproduce the problem, and we're happy to look into it. I initially didn't realise that this was just a copy of what was already there.

Reply:
👍 Reproduced it here https://github.com/DataDog/dd-trace-go/issues/849